### PR TITLE
[#59436] Set reminders for work packages to automatically generate date alert notifications (backend services)

### PR DESCRIPTION
**Pull request by akabiru**
# Ticket

https://community.openproject.org/work_packages/59436

# What are you trying to accomplish?

- [x] [Op#59433](https://community.openproject.org/projects/communicator-stream/work_packages/59433) Define Database Models for reminders 
- [x] [Op#59434](https://community.openproject.org/projects/communicator-stream/work_packages/59434)
   - [x] Add Notification reason `reminders`
   - [x] Visualize reminders in notification center https://github.com/opf/openproject/pull/17323
- [x] [OP#59435](https://community.openproject.org/projects/communicator-stream/work_packages/59435)
  - [x] Contracts
  - [x] Reminders create service - schedules notification creation at _`remind_at`_
    - [x] Create a "reminder_notification" - join table that tracks notifications created from reminders
  - [x] Reminders update service
  - [x] Reminders delete service
- [x] Handle seeding and migrating of new `manage_own_reminders` permission

## Screenshots

Screenshot 2024-12-02 at 5 45 26 PM

> [!NOTE]
> Reminder note covered in #17323 

# What approach did you choose and why?

Reminders are set as a polymorphic association to allow for any future extension to other models such as comments, activities or wiki pages (which are out of scope for this first version)

At present a user can only create their own reminders; if there is a need to create for other users, an explicit `recipient` can be added to the reminders table. Right now, we take a minimal approach.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

**Comment by akabiru** [comment]
@jjabari-op as discussed- reverted back to reminder state inference model. https://github.com/opf/openproject/pull/17229/commits/b9a5fd0cd5fe5d1560fd003e08ea0f56cc35d884

**Comment by akabiru** [comment]
Broken tests are from https://github.com/opf/openproject/pull/17267 - Markus will look into it

File: app/contracts/reminders/base_contract.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Reminders
+  class BaseContract < ::ModelContract
+    MAX_NOTE_CHARS_LENGTH = 80
```
Comment (jjabari-op) [inline app/contracts/reminders/base_contract.rb:31]: that seems very short

**Review by jjabari-op — APPROVED** [review]
Great work @akabiru!

The only thing I've found was the max length of the notes, which I think is too short. Should be discussed with Wieland probably.

As everything else looks good, I approve this PR!

File: app/contracts/reminders/base_contract.rb (hunk 1 above)
Comment (akabiru) [reply app/contracts/reminders/base_contract.rb:31]: Good eye! We discussed this actually and 80 chars [was the arrived spec](https://community.openproject.org/wp/59436). We can revisit and easily change if too short.

Screenshot 2024-12-04 at 5 46 12 PM

**Pull request merged**